content/advent-2016/peachpy.md

```diff
@@ -0,0 +1,377 @@
+arguments to the actual function stub, so a mismatch here won't be detected.
+
+[embedmd]:# (args/add.py /with.*/)
+```py
+with Function("add", (f1, f2), uint64_t, target=uarch.default) as function:
+```
+
+This line declares a function `add` with two arguments `(f1, f2)` and with a
+return value of type `uint64_t`.
+
+[embedmd]:# (args/add.py /.*reg_f1/ /(?m:^$)/)
+```py
+    reg_f1 = GeneralPurposeRegister64()
+    reg_f2 = GeneralPurposeRegister64()
+```
+
+Next, we acquire two registers.  In this case, we don't care what the actual
+registers are. PeachPy will assign two for us.
+
+[embedmd]:# (args/add.py /.*LOAD.*reg_f1/ /(?m:^$)/)
+```py
+    LOAD.ARGUMENT(reg_f1, f1)
+    LOAD.ARGUMENT(reg_f2, f2)
+```
+
+These are PeachPy pseudo-instructions which will load the arguments into the
+chosen registers.  This is where details are hidden allowing portable assembly.
+With the Go calling convention the arguments are passed on the stack, and so
+PeachPy generates the appropriate `MOV` instructions.  If PeachPy generates
+code for the regular C calling convention, those will load the arguments from
+the appropriate registers instead.
+
+[embedmd]:# (args/add.py /.*ADD.*/ /RETURN.*/ )
+```py
+    ADD(reg_f1, reg_f2)
+
+    RETURN(reg_f1)
+```
+
+This last block does the addition and then sets the return value.  Here again
+`RETURN` is a PeachPy pseudo-instruction which will generate the appropriate
+store into the stack for Go's calling convention.
+
+Note also that the operands are Intel syntax order (destination first), rather
+than the Plan9/AT&T order (source first).  PeachPy generates the correct
+assembly syntax depending on the target.
+
+We can then call this as a normal Go function.
+
+[embedmd]:# (args/main.go /func main/ /}/ )
+```go
+func main() {
+	sum := add(100, 20)
+	fmt.Printf("sum = %+v\n", sum)
+}
+```
+
+**Working with structs**
+
+This example will talk about passing a struct by reference to a function.
+
+First, in order to be able to access any of the fields, you need to know where
+their offset relative to the base address of the struct.  The `unsafe.Offsetof`
+function will do this for you, but so can Dominik Honnef's `structlayout` tool.
+
+For example, given
+
+[embedmd]:# (struct/main.go /type foo struct/ /}/)
+```go
+type foo struct {
+	zot uint16
+	bar uint16
+	qux uint64
+}
+```
+
+We can query the offsets of all the fields with:
+
+```
+    $ structlayout github.com/dgryski/peachpy-examples/struct foo
+    foo.zot uint16: 0-2 (size 2, align 2)
+    foo.bar uint16: 2-4 (size 2, align 2)
+    padding: 4-8 (size 4, align 0)
+    foo.qux uint64: 8-16 (size 8, align 8)
+```
+
+And here's our PeachPy code:
+
+[embedmd]:# (struct/add.py)
+```py
+import peachpy.x86_64
+
+f = Argument(ptr())
+
+with Function("add", (f,), uint64_t, target=uarch.default) as function:
+    reg_f_base = GeneralPurposeRegister64()
+
+    LOAD.ARGUMENT(reg_f_base, f)
+
+    v = GeneralPurposeRegister64()
+
+    MOVZX(v, word[reg_f_base+2])
```

Comment:
In Intel assembly `word` means 16-bits wide because that's how it was on old 16-bit systems (think Intel 286), and the convention was preserved for compatibility reasons into the modern 64-bit systems, but its not obvious :) In many other instruction sets `word` means either 32-bit or 64-bit wide.